sklearn/metrics/ranking.py

```diff
@@ -172,16 +190,48 @@ def average_precision_score(y_true, y_score, average="macro",
     >>> y_true = np.array([0, 0, 1, 1])
     >>> y_scores = np.array([0.1, 0.4, 0.35, 0.8])
     >>> average_precision_score(y_true, y_scores)  # doctest: +ELLIPSIS
-    0.79...
+    0.83...
+
+    >>> yt = np.array([0, 0, 1, 1])
+    >>> ys = np.array([0.1, 0.4, 0.35, 0.8])
+    >>> ap = average_precision_score(yt, ys, interpolation='eleven_point')
+    >>> ap # doctest: +ELLIPSIS
+    0.84...
 
     """
-    def _binary_average_precision(y_true, y_score, sample_weight=None):
+    def _binary_uninterpolated_average_precision(
+            y_true, y_score, sample_weight=None):
         precision, recall, thresholds = precision_recall_curve(
             y_true, y_score, sample_weight=sample_weight)
-        return auc(recall, precision)
+        # Return the step function integral
+        return -np.sum(np.diff(recall) * np.array(precision)[:-1])
 
-    return _average_binary_score(_binary_average_precision, y_true, y_score,
-                                 average, sample_weight=sample_weight)
+    def _binary_eleven_point_average_precision(
+            y_true, y_score, sample_weight=None):
+        precision, recall, thresholds = precision_recall_curve(
+            y_true, y_score, sample_weight=sample_weight)
+
+        # We need the recall values in ascending order, and to ignore the first
+        # (precision, recall) pair with precision = 1.
+        precision = precision[-2::-1]
+        recall = recall[-2::-1]
+
+        return np.mean([precision[i:].max() for i in
+                        np.searchsorted(recall, np.arange(0, 1.1, 0.1))])
```

Comment:
It would be good to move those two functions outside the enclosing function. In general, functions defined inside functions is not great style, as it hinders reuse and testing.

Reply:
@ndingwall I believe having them at the module level or at `metrics.base` would be preferred.